lhotse/dataset/speech_recognition.py

```diff
@@ -57,3 +58,108 @@ def __getitem__(self, idx: int) -> Dict[str, torch.Tensor]:
+        cut = self.cuts[cut_id]
+
+        features = torch.from_numpy(cut.load_features())
+
+        return {
+            'features': features,
+            'supervisions': [
+                {
+                    'example_idx': idx,
+                    'text': sup.text,
+                    'start_frame': round(sup.start / cut.frame_shift),
+                    'end_frame': round(sup.end / cut.frame_shift),
+                }
+                # CutSet's supervisions can exceed the cut, when the cut starts/ends in the middle
+                # of a supervision (they would have relative times e.g. -2 seconds start, meaning
+                # it started 2 seconds before the Cut starts). We use s.trim() to get rid of that
+                # property, ensuring the supervision time span does not exceed that of the cut.
+                for sup in (s.trim(cut.duration) for s in cut.supervisions)
+            ]
+        }
+
+    def __len__(self) -> int:
+        return len(self.cut_ids)
+
+
+class K2DataLoader(DataLoader):
+    """
+    A PyTorch DataLoader that has a custom collate_fn that complements the K2SpeechRecognitionDataset.
+
+    The 'features' tensor is collated in a standard way to return a tensor of shape (B, T, F).
+
+    The 'supervisions' dict contains the same fields as in ``K2SpeechRecognitionDataset``,
+    except that each sub-field (like 'start_frame') is a 1D PyTorch tensor with shape (B,).
+    The 'text' sub-field is an exception - it's a list of strings with length equal to batch size.
+
+    The 'example_idx' sub-field in 'supervisions', which originally points to index of the example
+    in the Dataset, is remapped to the index of the corresponding features matrix in the
+    collated 'features'.
+    Multiple supervisions coming from the same cut will share the same 'example_idx'.
+
+    For example, see ``test/dataset/test_speech_recognition_dataset.py::test_k2_dataloader()``.
+    """
+
+    def __init__(self, *args, **kwargs):
+        if 'collate_fn' in kwargs:
+            raise ValueError('Cannot override collate_fn in K2DataLoader.')
+        super().__init__(*args, collate_fn=multi_supervision_collate_fn, **kwargs)
+
+
+def multi_supervision_collate_fn(batch: Sequence[Dict]) -> Dict:
+    """Custom collate_fn for K2SpeechRecognitionDataset. See its docs for details."""
+    from torch.utils.data._utils.collate import default_collate
+
+    dataset_idx_to_batch_idx = {
+        example['supervisions'][0]['example_idx']: batch_idx
+        for batch_idx, example in enumerate(batch)
+    }
+
+    def update(d: Dict, **kwargs) -> Dict:
+        for key, value in kwargs.items():
+            d[key] = value
+        return d
+
+    supervisions = default_collate([
+        update(sup, example_idx=dataset_idx_to_batch_idx[sup['example_idx']])
+        for example in batch
+        for sup in example['supervisions']
+    ])
+    feats = default_collate([example['features'] for example in batch])
+    return {
```

Comment:
I wonder if you could explain what this function returns, in terms of tensors and their dims, or whatever?  It's not super easy to unpack from the code.

Reply:
The return value can be used in k2 to construct a dense fsa:

```python
def dense_fsa(log_probs: torch.Tensor,
              supervision_segments: torch.Tensor) -> _k2.DenseFsaVec:
    '''Construct a DenseFsaVec from neural net log-softmax outputs.
    Args:
      log_probs:
        A 3-D tensor of dtype ``torch.float32`` with shape ``(N, T, C)``, where
        ``N`` is the number of sequences, ``T`` the maximum input length,
        and ``C`` the number of output classes.
      supervision_segments:
        A 2-D CPU tensor of dtype ``torch.int32`` with 3 columns.
        Each row contains information for a supervision segment. Column 0 is the
        ``sequence_index`` indicating which sequence this segment comes from;
        column 1 specifies the ``start_frame`` of this segment within the
        sequence; column 2 contains the ``duration`` of this segment.
```

`feats` can be used as input to a neural network and its output is `log_probs`.

`supervisions` contains information for constructing `supervision_segments`:
 - column 0 of `supervision_segments` is from `supervisions.example_idx[i]`
 - column 1 is from `supervisions.start_frame[i]`
 - column 2 is from `supervisions.end_frame[i]`

`supervisions.text[i]` can be used to construct a linear FSA.

Reply:
Thanks!
I wonder whether it would make sense to have the collate_fn directly create the supervision_segments tensor of shape (num_supervisions,3).  Not sure if that's considered py-torchy or not.

Reply:
I think it is possible. The returned `example_idx`, `start_frame` and `end_frame` are just 1-D torch.Tensors, we
can `stack` them and make a transpose to get a tensor of shape `(num_supervisions, 3)`.

Reply:
I think it's fine, I'll make these changes. I'll rename example_idx to sequence_idx for consistency with K2 (given what I see in `dense_fsa()` doc-string). Also, one mismatch I noted is that I'm returning `end_frame`, and the doc provided by Fangjun says it expects `duration` - should I change this?

Reply:
> and the doc provided by Fangjun says it expects duration - should I change this

Yes, that would be nice.